File: typescript/examples/getting-started/07-embeddings.ts
// Embeddings via the embed() builtin
// Generate vector embeddings for text, useful for semantic search and similarity.
import { z } from "zod";
import { Opper } from "../../src/index.js";

const opper = new Opper();

// ── Single text embedding ───────────────────────────────────────────────────

const result = await opper.call("sdk-test-embed", {
  input_schema: z.object({
    text: z.string().describe("Text to generate an embedding for"),
  }),
  output_schema: z.object({
    embedding: z.array(z.number()).describe("Vector embedding of the input text"),
  }),
  input: {
    text: "The quick brown fox jumps over the lazy dog",
  },
});

console.log("── Single embedding ──");
console.log("Dimensions:", result.data.embedding.length);
console.log("First 5 values:", result.data.embedding.slice(0, 5));

// ── Compare similarity between two texts ────────────────────────────────────

const a = await opper.call("sdk-test-embed", {
  input_schema: z.object({ text: z.string().describe("Text to embed") }),
  output_schema: z.object({ embedding: z.array(z.number()).describe("Vector embedding") }),
  input: { text: "I love programming in TypeScript" },
});

const b = await opper.call("sdk-test-embed", {
  input_schema: z.object({ text: z.string().describe("Text to embed") }),
  output_schema: z.object({ embedding: z.array(z.number()).describe("Vector embedding") }),
  input: { text: "TypeScript is my favorite language for coding" },
});

const c = await opper.call("sdk-test-embed", {
  input_schema: z.object({ text: z.string().describe("Text to embed") }),
  output_schema: z.object({ embedding: z.array(z.number()).describe("Vector embedding") }),
  input: { text: "The weather in Stockholm is cold today" },
});

function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0, magA = 0, magB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    magA += a[i] * a[i];
    magB += b[i] * b[i];
  }
  return dot / (Math.sqrt(magA) * Math.sqrt(magB));
}

console.log("\n── Similarity comparison ──");
console.log("Similar texts:", cosineSimilarity(a.data.embedding, b.data.embedding).toFixed(4));
console.log("Unrelated texts:", cosineSimilarity(a.data.embedding, c.data.embedding).toFixed(4));

File: python/examples/getting-started/07_embeddings.py
# Embeddings: generate vector embeddings for semantic similarity
import math

from opperai import Opper

opper = Opper()


# ── Single text embedding ───────────────────────────────────────────────────

result = opper.call(
    "sdk-test-embed",
    input={"text": "The quick brown fox jumps over the lazy dog"},
    input_schema={"type": "object", "properties": {"text": {"type": "string"}}},
    output_schema={
        "type": "object",
        "properties": {"embedding": {"type": "array", "items": {"type": "number"}}},
    },
    model="openai/text-embedding-3-small",
)

print("── Single embedding ──")
print("Dimensions:", len(result.data["embedding"]))
print("First 5 values:", result.data["embedding"][:5])


# ── Compare similarity between two texts ────────────────────────────────────


def embed(text: str) -> list[float]:
    r = opper.call(
        "sdk-test-embed",
        input={"text": text},
        input_schema={"type": "object", "properties": {"text": {"type": "string"}}},
        output_schema={"type": "object", "properties": {"embedding": {"type": "array", "items": {"type": "number"}}}},
        model="openai/text-embedding-3-small",
    )
    return r.data["embedding"]


def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    mag_a = math.sqrt(sum(x * x for x in a))
    mag_b = math.sqrt(sum(x * x for x in b))
    return dot / (mag_a * mag_b)


vec_a = embed("I love programming in Python")
vec_b = embed("Python is my favorite language for coding")
vec_c = embed("The weather in Stockholm is cold today")

print("\n── Similarity comparison ──")
print(f"Similar texts: {cosine_similarity(vec_a, vec_b):.4f}")
print(f"Unrelated texts: {cosine_similarity(vec_a, vec_c):.4f}")
